### ISSUE TITLE
Already-failed Deferred doesn't re-raise exception when awaited

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8878]
|[<img alt="IlyaSkriblovsky's avatar" src="https://avatars.githubusercontent.com/u/909638?s=50" width="50" height="50">](https://github.com/IlyaSkriblovsky)| @IlyaSkriblovsky reported|
|-|-|
|Trac ID|trac#8878|
|Type|defect|
|Created|2016-10-29 16:55:00Z|
|Branch|https://github.com/twisted/twisted/tree/8878-already-failed-await|

When I `await` a Deferred that is already failed, the exception isn't propagated, but it is instead returned like a normal result. The following code was tested with current Twisted trunk. In `foo()` exception is thrown as expected, but in `bar()` it is just returned which is inconsistent and is probably a bug.

```python
from twisted.internet.task import react
from twisted.internet import defer


async def foo(reactor):
    print('foo')
    d = defer.Deferred()
    reactor.callLater(0.1, d.errback, Exception('BOOOM'))
    try:
        result = await d
    except Exception as e:
        print('thrown', e)
    else:
        print('result', result)

async def bar():
    print('bar')
    try:
        result = await defer.fail(Exception('BOOOM'))
    except Exception as e:
        print('thrown', e)
    else:
        print('result', result)


@defer.inlineCallbacks
def main(reactor):
    yield defer.ensureDeferred(foo(reactor))
    yield defer.ensureDeferred(bar())

react(main)
```

<details><summary>Searchable metadata</summary>

```
trac-id__8878 8878
type__defect defect
reporter__IlyaSkriblovsky IlyaSkriblovsky
priority__normal normal
milestone__Twisted_16_6 Twisted 16.6
branch__8878_already_failed_await 8878-already-failed-await
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1477760100055257 1477760100055257
changetime__1478131670104176 1478131670104176
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl set owner to @hawkowl|
|-|-|

Yeah, confirmed. I have a fix, stand by...

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl removed owner|
|-|-|

PR at https://github.com/twisted/twisted/pull/577, please review

**Comment by twisted-trac** [comment]
|[<img alt="IlyaSkriblovsky's avatar" src="https://avatars.githubusercontent.com/u/909638?s=50" width="50" height="50">](https://github.com/IlyaSkriblovsky)<a name="note_4"></a>|@IlyaSkriblovsky commented|
|-|-|

Thanks, my snipped works fine with your PR

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_5"></a>|@moshez set owner to @hawkowl|
|-|-|

Reviewed on GH. Removed review kw, reassigning to Hawkie.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl set status to `closed`|
|-|-|

In changeset 737709b42c5e74faf1f38d14d4390a2075af4730
```
#!CommitTicketReference repository="" revision="737709b42c5e74faf1f38d14d4390a2075af4730"
Merge 8878-already-failed-await: Fix resolved Deferreds not raising up exceptions when awaited

Author: hawkowl
Reviewer: moshez
Fixes: #8878
```
